adaptdl/adaptdl/goodput.py

```diff
@@ -0,0 +1,256 @@
+        params[2] = upper[2] = lower[2]
+        params[3] = upper[3] = lower[3]
+    if not any(np.logical_and(nodes == 1, replicas > 1)):
+        # Fix alpha_r and beta_r if no single-node/multi-replica observations.
+        params[4] = upper[4] = lower[4]
+        params[5] = upper[5] = lower[5]
+    if not any(replicas > 2):
+        # Fix beta_n and beta_r if no replicas > 2.
+        params[3] = upper[3] = lower[3]
+        params[5] = upper[5] = lower[5]
+    bounds = scipy.optimize.Bounds(lower, upper, keep_feasible=True)
+    args = (nodes, replicas, local_bsz, accumulation_steps,
+            step_time, step_time_compute, accumulation_time)
+    # FIXME: need to handle optimization failures and propagate to the Trainer.
+    grad_fn = autograd.grad(_obj_fn)
+    result = scipy.optimize.minimize(_obj_fn, params, args=args,
+                                     jac=grad_fn, bounds=bounds)
+    params = result.x
+    np = orig_np  # Restore original numpy.
+    return Params(*params)
+
+
+def _predict(params, nodes, replicas, local_bsz, accumulation_steps):
+    params = Params(*params)
+    step_time_compute = _predict_compute(params, local_bsz)
+    step_time_network = _predict_network(params, nodes, replicas)
+    step_time_accumulation = step_time_compute
+    gamma = params.gamma
+    # Return predicted total step time in log-space to avoid numerical issues
+    # in autograd and optimization.
+    step_time = (
+        ((step_time_compute) ** gamma + step_time_network ** gamma) **
+        (1 / gamma)
+        + step_time_accumulation * accumulation_steps)
+    return (
+        step_time,
+        step_time_compute,
+        step_time_network)
+
+
+def _predict_compute(params, local_bsz):
+    params = Params(*params)
+    # Forward/backward passes should scale linearly with the batch size.
+    return params.alpha_c + params.beta_c * local_bsz
+
+
+def _predict_network(params, nodes, replicas):
+    params = Params(*params)
+    # Select the most significant link between replicas, currently either
+    # inter-node (nodes > 1) or intra-node (replicas > 1). Note that if
+    # replicas == 1 then neither of these two conditions are matched.
+    conds = [nodes > 1, replicas > 1]
+    # Bandwidth is bottlenecked by the most significant link, alpha models
+    # the overhead of transferring data across that link.
+    bottleneck = np.select(conds, [params.alpha_n, params.alpha_r], 1e-8)
+    # Assuming ring all-reduce, communication happens in a number of rounds
+    # equal to the number of replicas. beta models the performance
+    # retrogression from increasing the number of replicas beyond 2.
+    retrogress = np.select(conds, [params.beta_n, params.beta_r], 1e-8)
+    retrogress = retrogress * np.maximum(replicas - 2, 1e-8)
+    return (bottleneck + retrogress)
+
+
+def _rmse(pred, true):
+    return np.sqrt(((pred - true) ** 2).mean())
+
+
+def _obj_fn(params, nodes, replicas, local_bsz, accumulation_steps,
```

Comment:
Nit: I think the file would read better if `_obj_fn` was closer to `fit_perf_params`